storage-node/src/services/sync/acceptPendingObjects.ts

```diff
@@ -88,29 +88,25 @@ export class AcceptPendingObjectsService {
   }
 
   private async processPendingObjects(pendingIds: string[]): Promise<PendingObjectDetails> {
-    const pendingDataObjects = await this.qnApi.getDataObjectsByIds(pendingIds)
+    const pendingDataObjects = await this.qnApi.getDataObjectDetails(pendingIds)
 
     // objects not found in the query node
-    const maybeDeletedObjects = _.differenceWith(
-      pendingIds,
-      pendingDataObjects,
-      (id, dataObject) => dataObject.id === id
+    const maybeDeletedObjectIds = pendingIds.filter(
+      (id) => !pendingDataObjects.some((dataObject) => dataObject.id === id)
     )
 
-    if (maybeDeletedObjects.length) {
-      // unlink the pending object once we determined it is gone!
-      logger.debug(`Found ${maybeDeletedObjects.length} stale pending objects in pending folder`)
-      // Things to consider.
-      // These may be very recently uploaded files, note the uploads api uses chain rpc to
-      // validate upload of object that was created to allow uploads for objects that were created
-      // as soon as possible.
-      // const qnState = await this.qnApi.getQueryNodeState()
-      // get this.api chain head
-      // get timestamp in chain head block.. are we uptodate?
-      // test api.isSyncing?
-      // It would be easier to have query-node store object deleted events.
-      // if we chose a different file name convention for the pending objects,
-      // we could store the bag-id and object id in the filename to make it easy to lookup in runtime state.
+    if (maybeDeletedObjectIds.length) {
+      logger.debug(`Found ${maybeDeletedObjectIds.length} stale pending objects in pending folder`)
+
+      // Check if the stale objects have been ACTUALLY deleted using the deleted events from Storage Squid
+      const deletedObjectIds = await this.qnApi.getDataObjectDeletedEvents(maybeDeletedObjectIds)
+
+      // Delete the stale objects from the pending folder
+      await Promise.allSettled(
+        deletedObjectIds.map(({ data: { dataObjectId } }) =>
+          fsPromises.unlink(path.join(this.pendingDataObjectsDir, dataObjectId))
```

Comment:
If an object has been deleted from the runtime then, `pendingDataObjects` wont contain it in the first place since QN wont return it. Wouldn't that be the case?

Reply:
we might actually try to compute the ipfs hash of it first.. but still the general point applies. correct?

Reply:
Sorry my bad, I confused it with the `pendingIds` array.